query/data_filter.go

```diff
@@ -0,0 +1,128 @@
+// Copyright 2023 Searis AS
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package query
+
+import (
+	"slices"
+	"time"
+)
+
+// DataFilter describe a type can return an internal data filter structure.
+// Data filters helps reduce the amount of data that is returned by a method.
+type DataFilter interface {
+	filter() dataFilter
+}
```

Comment:
I believe DataFilter is an internal interface, so we can use `dataFilter`

Reply:
?

Reply:
I mean the interface is public, while its signature is a private function. So I'm curious if it's done with the intention of forbidding users from adding their customer implementations of `DataFilter` or why?

Reply:
Right.

So, this is a _public_ interface with an _unexported_ method, and all it's side effects are intentional.

That means that:
1. We can recognize valid filter implementations.
2. Users can't implement their own filter implementations.
3. We don't need to maintain any _backwards compatibility_ (after the 1.0 release) regarding the internal dataFilter _structure_. That's the main advantage of this approach.

Reply:
You mention that we could make the `DataFilter` interface itself private as well, and yes, we could, but that makes usage akward. E.g. you may want to be able to do:

```go
var dq DataFilter
if x {
    dq = ...
}
```

If the interface type is private, we can't do that.

Reply:
Apparently it's called a "Sealed Interface".

https://blog.chewxy.com/2018/03/18/golang-interfaces/

Reply:
Example of sealed interfaces can be found in the `otel` pacakge. E.g.:

https://pkg.go.dev/go.opentelemetry.io/otel/metric@v1.19.0/embedded#Meter

Reply:
It's interesting that GoDoc even comments that interface contains private functions 
`// contains filtered or unexported methods`

Reply:
It seams the "common" way of doing sealed interfaces is actually to do:

```go
type I interface{
  method()
}
```

With no return value. I suppose this means they do a _type switch_ where the I type is accepted and handle it appropriately. We _could_ do that as well, later on.

Reply:
Actually, may need to switch from interface to using a strict type... We have examples that tries to json.Unmarshal stuff into filters; need to make sure this still works!